### ISSUE TITLE
Strange behavior of sol::overload

### ISSUE THREAD

**Issue by Ung0d** [ThePhD/sol2#136]
Hey. I want to bind a set of overloaded c++ methods to a set of overloaded lua methods.
However, running this code yields in some kind of permaloop and the program has to be shut down.

```
struct TestMe
{
    void print(int i) { std::cout << "Integer print: " << i << std::endl; }
    void print() { std::cout << "No param print." << std::endl; }
};

int main()
{
    auto test = std::make_unique<TestMe>();
    sol::state lua;
    lua.new_usertype<TestMe>("TestMe", sol::constructors<>(),
                              "print", sol::overload(static_cast<void (TestMe::*)(int)>(&TestMe::print),
                                                     static_cast<void (TestMe::*)()>(&TestMe::print)));
    lua.set("test", std::move(test));

    sol::protected_function protFunc = lua.load("test:print()");
    auto res = protFunc();
    if(!res.valid())
    {
        sol::error err = res;
        std::cout << "Something went wrong in the execution of a script!" << std::endl << "What: " << err.what() << std::endl;
    }

    return 0;
}
```

Note that you can swap the arguments in sol::overload and it works fine:

```
    lua.new_usertype<TestMe>("TestMe", sol::constructors<>(),
                              "print", sol::overload(static_cast<void (TestMe::*)()>(&TestMe::print),
                                                     static_cast<void (TestMe::*)(int)>(&TestMe::print)));
```

Is there a 'order restriction' when using sol:overload? It seems like sol dont like it when methods with no param are registered after methods with one or more param.

**Comment by ThePhD** [comment]
No repro with lua 5.3.2/5.3.3 and VC++. Which compiler, platform, build flags, etc...

**Comment by Ung0d** [comment]
Gcc 5.3, Win7 64bit, occurs even with -std=c++14 only.

**Comment by ThePhD** [comment]
Found the bug that's causing it to not find the right function, but I haven't gotten it to go into an infinite loop, ordered in any way. Will have to try with MinGW sometime.

**Comment by Ung0d** [comment]
Well I cant describe it better then that: Program simply does not terminate when encountering this issue. Just displaying empty terminal. What is your output? An error? Just false print call? I have the feeling that generally calling a (member-) method with wrong argument count in lua - not especially overloaded ones - triggers that behavior for me. So it seems like that "infinite-loop" does not directly related to the overload issue and is more a "function with wrong argument number called"-thing. 

Basically your bugfix should at least fix my permaloop in this situation.

**Comment by ThePhD** [comment]
@Ung0d Give it a go now with the latest commit, you should be all set.
